File: julia/day01.jl
module Day01

function parse_input()
    path = joinpath(@__DIR__, "..", "golang", "day01", "input.txt")
    input = open(path, "r") do file
        read(file, String)
    end
    sonar_reports = parse.(Int, split(input))
    return sonar_reports
end

function part1()
    ssrs = parse_input()
    vals = zip(ssrs[1:end-1], ssrs[2:end])
    return count(v -> (v[1] < v[2]), vals)
end

function part2()
    ssrs = parse_input()
    triples = collect(zip(ssrs[1:end-2], ssrs[2:end-1], ssrs[3:end]))
    vals = zip(triples[1:end-1], triples[2:end])
    return count(v -> (v[1][1] + v[1][2] + v[1][3] < v[2][1] + v[2][2] + v[2][3]), vals)
end

end # module

println(Day01.part1())
println(Day01.part2())

File: python/day01.py
#!/usr/bin/env python

import argparse
import os.path


def parse(s: str) -> list[int]:
    return [int(x) for x in s.split()]


def part1(s: str) -> int:
    numbers = parse(s)
    return sum(numbers[i] > numbers[i - 1] for i in range(1, len(numbers)))


def part2(s: str) -> int:
    numbers = parse(s)
    return sum(numbers[i] > numbers[i - 3] for i in range(3, len(numbers)))


EXAMPLE = """\
199
200
208
210
200
207
240
269
260
263
"""


def test_parse() -> None:
    assert parse(EXAMPLE) == [199, 200, 208, 210, 200, 207, 240, 269, 260, 263]


def test_part1() -> None:
    assert part1(EXAMPLE) == 7


def test_part2() -> None:
    assert part2(EXAMPLE) == 5


def main() -> int:
    fname = os.path.join(
        os.path.dirname(__file__),
        "data",
        os.path.splitext(os.path.basename(__file__))[0] + ".txt",
    )

    with open(fname) as f:
        input = f.read()
        print(part1(input))
        print(part2(input))

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
